R4R: Multisig support (#3264)
cosmos/cosmos-sdk: docs/gaia/gaiacli.md

We strongly recommend _NOT_ using the same passphrase for multiple keys. The Tendermint team and the Interchain Foundation will not be responsible for the loss of funds.
:::

#### Generate multisig public keys

You can generate and print a multisig public key by typing:

```bash
gaiacli keys add --multisig=name1,name2,name3[...] --multisig-threshold=K new_key_name
```

`K` is the minimum number of private keys that must have signed the
transactions that carry the public key's address as signer.

The `--multisig` flag must contain the name of public keys that will be combined into a
public key that will be generated and stored as `new_key_name` in the local database. All names
supplied through `--multisig` must already exist in the local database. Unless the flag
`--nosort` is set, the order in which the keys are supplied on the command line does not matter

> as they are sorted in lexicoprapical order before the new key is generated is on, i.e. the

Reviewer: ```suggestion
as they are sorted in lexicographical order before the new key is generated, i.e. the
```

We strongly recommend _NOT_ using the same passphrase for multiple keys. The Tendermint team and the Interchain Foundation will not be responsible for the loss of funds.
:::

#### Generate multisig public keys

You can generate and print a multisig public key by typing:

```bash
gaiacli keys add --multisig=name1,name2,name3[...] --multisig-threshold=K new_key_name
```

`K` is the minimum number of private keys that must have signed the
transactions that carry the public key's address as signer.

> The `--multisig` flag must contain the name of private or public keys that will be combined into a

Reviewer 2: ```suggestion
The `--multisig` flag must contain the name of public keys that will be combined into a
```

Why would we need the private keys?

public key that will be generated and stored as `new_key_name` in the local database. All names
supplied through `--multisig` must already exist in the local database. The order in which the keys
are supplied on the command line matters, i.e. the following commands generate two distinct
multisig keys:

```bash
gaiacli keys add --multisig=foo,bar,baz --multisig-threshold=2 multisigaddress
```

```bash
gaiacli keys add --multisig=baz,foo,bar --multisig-threshold=2 multisigaddress
```

> Although keys and threshold parameter used in the above examples are the same, the commands

Reviewer 2: Can we sort the names? This behaviour seems unusual, unless there's a specific reason we want it.

Reviewer 2: ... actually, can we sort by the public keys? that way different names on different machines will still work

Reviewer 3: This sounds like a good idea. Ideally order doesn't matter here.

Reviewer: yeah I second this. I can see simple mishaps happening due to this.

Author: Sorted - sorry for the pun.

Author: I've also added the flag `--nosort` so that a user can pass the keys in the exact order he wants and expect they don't get shuffled.

generate two distinct keys due to the order in which the keys are specified.

> Multisig addresses can also be generated on-the-fly and printed through the which command:

Reviewer 3: This needs a bit more explanation, or should be 🔪

Author: Done

gaiacli query gov param tallying
gaiacli query gov param deposit
```

### Multisig transactions

Multisig transactions require signatures of multiple private keys. Thus, generating and signing
a transaction from a multisig account involve cooperation among the parties involved. A multisig

> transaction can be initiated by any of the keys holders, and at least one of them would need to

Reviewer 2: ```suggestion
transaction can be initiated by any of the key holders, and at least one of them would need to
```

import other parties' public keys into their local database and generate a multisig public key
in order to finalize and broadcast the transaction.

For example, given a multisig key comprising the keys `p1`, `p1`, and `p3`, each of which is held
by a distinct party, the user holding `p1` would require to import both `p2` and `p3` in order to
generate the multisig account public key:

```
gaiacli keys add \
  --pubkey=cosmospub1addwnpepqtd28uwa0yxtwal5223qqr5aqf5y57tc7kk7z8qd4zplrdlk5ez5kdnlrj4 \
  p2

gaiacli keys add \
  --pubkey=cosmospub1addwnpepqgj04jpm9wrdml5qnss9kjxkmxzywuklnkj0g3a3f8l5wx9z4ennz84ym5t \
  p3

gaiacli keys add \

>   --multisig=2

Reviewer 3: ```suggestion
  --multisig-threshold=2
```

Author: Thanks!

>   --multisig=p1,p2,p3  # note: the order matters \

Reviewer 2: See above comment on order.

  p1p2p3
```

A new multisig public key `p1p2p3` has been stored, and its address will be
used as signer of multisig transactions:

```bash
gaiacli keys show --address p1p2p3
```

The first step to create a multisig transaction is to initiate it on behalf
of the multisig address created above:

```bash

> gaiacli tx send \

Reviewer: Just curious, why do we not support `--output-document` for `tx send` instead of piping?

Reviewer 3: I like this too.

Author: I agree. I'd propose to promote `flagOutfile` to be added to all post commands - although I'd do such change in an ensuing PR

  --from=<multisig_address> \
  --to=cosmos1570v2fq3twt0f0x02vhxpuzc9jc4yl30q2qned \
  --amount=10stake \
  --generate-only > unsignedTx.json
```

The file `unsignedTx.json` contains the unsigned transaction encoded in JSON.
`p1` can now sign the transaction with its own private key:

```bash
gaiacli tx sign \
  --multisig=<multisig_address> \
  --name=p1 \
  --output-document=p1signature.json \
  unsignedTx.json
```

Once the signature is generated, `p1` transmits both `unsignedTx.json` and
`p1signature.json` to `p2` or `p3`, which in turn will generate their
respective signature:

```bash
gaiacli tx sign \
  --multisig=<multisig_address> \
  --name=p2 \
  --output-document=p2signature.json \
  unsignedTx.json
```

`p1p2p3` is a 2-of-3 multisig key, therefore one additional signature

> is sufficient. `p2` can generate the multisig transaction by combining

Reviewer 2: Why `p2` specifically? Couldn't anyone run this (if they had the signature files)?

Author: Yes, of course. Anyone could. Amended accordingly.
